### PR TITLE
SPR1-2373: Flag MK+ visibility values of -2**31 as DATA_LOST

### PR DESCRIPTION
**Pull request by ludwigschwardt**
The next-generation correlator for MK+ indicates missing F-engine heaps by setting the corresponding correlation products in the X-engine to `-2 ** 31 + 1j`, whereas the old MK correlator set those to `0 + 0j`. Flag these values as `DATA_LOST`, which is more specific than the `CAM` flag used to flag zeroes in general.

The special value of `-2 ** 31` (aka the *missing data marker*) is unambiguous since it is outside the valid range of values in both the old and new correlators. The new correlator prefers this value to 0 because that makes 0 a valid value that can be used to check corner cases in its unit tests.

For now we keep on flagging zeros as `CAM` in MK+, since this would ensure continuity and still be an unlikely value in proper science operations. We also zero visibilities with the special value after flagging, in case it leaks through to folks who disregard flags, and flag it as `CAM` as well for good measure.

I used the opportunity to fix some technical debt too... There is a large overhead to PRs and by the time I get around to a technical debt PR, I would have to rediscover what I now know.

I now use negative zeros to distinguish missing data from plain zeros in the intermediate visibility product in the ingest pipeline, which maintains backward compatibility while also allowing more specific flagging.

### CODE REVIEW

**Comment by ludwigschwardt** [comment]
I've asked @bmerry multiple questions in SPR1-2373 and NGC-667. We should probably resolve those before accepting this PR...

File: katsdpingest/ingest_kernels/prepare_flags.mako [hunk 1]
```diff
@@ -34,6 +35,8 @@ KERNEL REQD_WORK_GROUP_SIZE(${block}, ${block}, 1) void prepare_flags(
         float2 v = vis[${r} * vis_stride + ${c}];
         if (v.x == 0 && v.y == 0)
             f |= zero_flag;
+        else if (v.x == -2147483600.0f)   // -2 ** 31 as a float32
```
Comment (bmerry) [inline katsdpingest/ingest_kernels/prepare_flags.mako]: This smells wrong: v.x will be the raw (int) value * (1/n_accs). It's possible your unit tests are passing only because they don't set n_accs (it defaults to 1).

Comment (ludwigschwardt) [reply katsdpingest/ingest_kernels/prepare_flags.mako]: Good point! I now check the int value in `prepare` instead.

**Comment by ludwigschwardt** [comment]
Ready for review again after the third redesign 😅

I would review the first commit (technical debt) and then all changes as a whole, to avoid the mess in the middle.

File: katsdpingest/ingest_kernels/prepare.mako [hunk 2]
```diff
@@ -19,12 +19,23 @@ KERNEL REQD_WORK_GROUP_SIZE(${block}, ${block}, 1) void prepare(
...
+     * Convert these to negative zero floats instead. This treats missing visibilities
+     * as zeros like in the old CBF, while still allowing them to be flagged
+     * differently in prepare_flags by checking their sign bits.
      */
     <%transpose:transpose_load coords="coords" block="${block}" vtx="${vtx}" vty="${vty}" args="r, c, lr, lc">
         int2 in_value = vis_in[${r} * vis_in_stride + ${c}];
         float2 scaled_value;
-        scaled_value.x = (float) in_value.x * scale;
-        scaled_value.y = (float) in_value.y * scale;
+        if (in_value.x == -1 << 31)
```
Comment (richarms) [inline katsdpingest/ingest_kernels/prepare.mako]: I wonder if using `(-2147483647 - 1)` as the comparison constant here instead of left-shifting a negative integer would be better. One could also consider `INT_MIN` if we're confident there is support through newer versions of OpenCL and CUDA C. I tested both and tests all pass.

**Review by richarms — APPROVED** [review]
This all looks good to me Ludwig. Read the commit history then reviewed this as a single diff and it's a coherent and elegant story. Thanks for updating and reworking the missing-data flag handling.

File: katsdpingest/ingest_kernels/prepare.mako (hunk 2 above)
Comment (ludwigschwardt) [reply katsdpingest/ingest_kernels/prepare.mako]: I chose `-1 << 31` because it is easy to check that it is correct. The shift operation happens compile time, so the emitted code should be identical. Compare the output of

https://godbolt.org/z/nMEf9cnxz  (-1 << 31)

with

https://godbolt.org/z/xd7W5xE19  (-2147483647 - 1)

On `INT_MIN:` I'm happy to use that if it is compatible with the spec. There I have questions for @bmerry like

- Did you intend it to be `INT_MIN`?
- What if `INT_MIN` ever changes (unlikely, I know 😋)?

Comment (bmerry) [reply katsdpingest/ingest_kernels/prepare.mako]: > What if INT_MIN ever changes (unlikely, I know 😋)?

If it does, that means the `int` type has changed too, in which case you're in much bigger trouble. Historically CUDA has not been great at supporting types like int32_t (in particular, I don't think there is a 2 x int32_t vector type), which is why we use base types like `int` even though the C/C++ standards don't define their sizes (at least C++20 finally required twos complement). On the other hand, I think that means there is so much installed CUDA code that depends on `int` being 32-bit that they're unlikely to ever change it.

Comment (ludwigschwardt) [reply katsdpingest/ingest_kernels/prepare.mako]: Would you recommend `INT_MIN` or `-1 << 31` then? I would need a matching constant on the Python side too.

Comment (bmerry) [reply katsdpingest/ingest_kernels/prepare.mako]: It looks like left shift of an unsigned value is undefined in C++17 (it's defined in C++20), so I think either INT_MIN or (-2147483647 - 1) may be ideologically a little safer. It looks like OpenCL C actually [defines](https://registry.khronos.org/OpenCL/specs/unified/html/OpenCL_C.html#integer-macros) INT_MIN to be (-2147483647 - 1).

Under CUDA I *think* INT_MIN is probably going to come from system headers, rather being something baked into CUDA. I see port.mako from katsdpsigproc includes math.h, float.h and stdio.h, none of which would explicitly define INT_MIN, but they may happen to be including another header that does. So that's a bit risky, and if you want to use INT_MIN I'd suggest adding an include of <limits.h> to port.mako (probably a good idea anyway to get consistency).

Comment (ludwigschwardt) [reply katsdpingest/ingest_kernels/prepare.mako]: Thanks Bruce!

> It looks like left shift of an unsigned value is undefined in C++17 ...

-1 is signed, right? Oh, I see, the issue is exactly with signed ints. Makes sense.

> INT_MIN is probably going to come from system headers...

Indeed, it comes from `limits.h`. I'll add that to katsdpsigproc then.

You mention C++17 / C++20 but also use C-style includes like `math.h` and `limits.h`. Out of interest, are the mako's compiled as C or C++?

**Comment by ludwigschwardt** [comment]
I've moved over to INT_MIN. It works out of the box, even before I added `limits.h` to `port.mako` 🤔 I'll take it, though, since that means we won't need to wait for a `katsdpsigproc` release 😁

I've also suppressed an NVCC compiler warning (although the warning only pops up the first time the ingest pipeline starts up, since the compiled code is cached in `~/.cache/pycuda/compiler-cache-v1/`).
